lenskit-slim/src/main/java/org/lenskit/slim/SLIMBuildContextProvider.java

```diff
@@ -0,0 +1,301 @@
+            if (logger.isTraceEnabled()) {
+                logger.trace("computing similarities for item {} ({} of {})",
+                        itemId1, ndone, nitems);
+            }
+            Long2DoubleSortedMap vec1 = LongUtils.frozenMap(itemVectors.get(itemId1));
+            if (vec1.size() < minCommonUsers) {
+                // if it doesn't have enough users, it can't have enough common users
+                if (logger.isTraceEnabled()) {
+                    logger.trace("item {} has {} (< {}) users, skipping", itemId1, vec1.size(), minCommonUsers);
+                }
+                progress.advance();
+                continue OUTER;
+            }
+
+            LongIterator itemIter = itemSimilarity.isSymmetric() ? allItems.iterator(itemId1) : allItems.iterator();
+
+            Long2DoubleAccumulator row = rows.get(itemId1);
+            INNER: while (itemIter.hasNext()) {
+                long itemId2 = itemIter.nextLong();
+                if (itemId1 != itemId2) {
+                    Long2DoubleSortedMap vec2 = LongUtils.frozenMap(itemVectors.get(itemId2));
+                    if (!LongUtils.hasNCommonItems(vec1.keySet(), vec2.keySet(), minCommonUsers)) {
+                        // items have insufficient users in common, skip them
+                        continue INNER;
+                    }
+
+                    double sim = itemSimilarity.similarity(vec1, vec2);
+                    if (threshold.retain(sim)) {
+                        row.put(itemId2, sim);
+                        npairs += 1;
+                        if (itemSimilarity.isSymmetric()) {
+                            rows.get(itemId2).put(itemId1, sim);
+                            npairs += 1;
+                        }
+                    }
+                }
+            }
+
+            progress.advance();
+        }
+        progress.finish();
+        logger.info("built model of {} similarities for {} items in {}",
+                npairs, ndone, progress.elapsedTime());
+
+        Long2ObjectMap<LongSet> neighborSets = finishRows(rows);
+
+        //frozen each set and fill up itemNeighbors
+        Iterator<Map.Entry<Long,LongSet>> itemNeighborIter = neighborSets.entrySet().iterator();
+        while (itemNeighborIter.hasNext()) {
+            Map.Entry<Long,LongSet> entry = itemNeighborIter.next();
+            long item = entry.getKey();
+            LongSortedSet value = LongUtils.frozenSet(entry.getValue());
+            itemNeighbors.put(item, value);
+            itemNeighborIter.remove();
+        }
+    }
+
+    private Long2ObjectMap<Long2DoubleAccumulator> makeAccumulators(LongSet items) {
+        Long2ObjectMap<Long2DoubleAccumulator> rows = new Long2ObjectOpenHashMap<>(items.size());
+        LongIterator iter = items.iterator();
+        while (iter.hasNext()) {
+            long item = iter.nextLong();
+            Long2DoubleAccumulator accum;
+            if (modelSize == 0) {

```

Comment:
Ok, so that's what you were doing with the `ItemItemModel` in the old code. I did not understand this.

It looks like, to fix that, we now have a copy of item-item's model-build here.

I propose: we have two different context providers (with appropriate code sharing), one for fsSLIM and one for the other one. The fsSLIM can use an item-item model, as that will have the cosine similarities pre-computed.